File: e-scala/src/main/scala/dev/akif/e/scala/JsonStringEncoder.scala
package dev.akif.e.scala

import dev.akif.e.AbstractJsonStringEncoder

object JsonStringEncoder extends AbstractJsonStringEncoder[Option[Throwable], Map[String, String]] {
  override protected def encodeCause(cause: Option[Throwable]): String =
    cause.fold("null")(c => s""""${escape(c.getMessage)}"""")

  override protected def encodeData(data: Map[String, String]): String =
    data.map { case (k, v) => s""""${escape(k)}":"${escape(v)}"""" }.mkString("{", ",", "}")
}

File: e-kotlin/src/main/kotlin/dev/akif/e/kotlin/JsonStringEncoder.kt
package dev.akif.e.kotlin

import dev.akif.e.AbstractJsonStringEncoder

object JsonStringEncoder : AbstractJsonStringEncoder<Throwable?, Map<String, String>>() {
    override fun encodeCause(cause: Throwable?): String =
        if (cause == null) "null" else """"${escape(cause.message)}""""

    override fun encodeData(data: Map<String, String>): String =
        data.map { e -> """"${escape(e.key)}":"${escape(e.value)}"""" }.joinToString(",", "{", "}")
}
